models/service_test.go

```diff
@@ -0,0 +1,205 @@
+	mockIstioClient := new(IstioClientMock)
+
+	serviceDetails := fakeServiceDetails()
+	mockIstioClient.On("GetServiceDetails", "namespace", "service").Return(serviceDetails, nil)
+
+	istioDetails := fakeIstioDetails()
+	mockIstioClient.On("GetIstioDetails", "namespace", "service").Return(istioDetails, nil)
+
+	return mockIstioClient
+}
+
+func mockPrometheusDetails() *prometheusClientMock {
+	mockPrometheusClient := new(prometheusClientMock)
+	prometheusDetails := fakePrometheusDetails()
+	mockPrometheusClient.On("GetSourceServices", "namespace", "service").Return(prometheusDetails, nil)
+
+	return mockPrometheusClient
+}
+
+func TestServiceDetailParsing(t *testing.T) {
+	mockIstioClient := mockServiceDetails()
+	mockPrometheusClient := mockPrometheusDetails()
+
+	service := models.Service{}
+	service.Name = "service"
+	service.Namespace = models.Namespace{"namespace"}
+	service.SetIstioClient(mockIstioClient)
+	service.SetPrometheusClient(mockPrometheusClient)
+	err := service.SetServiceDetails()
+	if err != nil {
+		assert.Equal(t, err, nil, "Error found during parsing")
+	}
+
+	mockIstioClient.AssertExpectations(t)
+	mockPrometheusClient.AssertExpectations(t)
+
+	// Kubernetes Details
+	assert.Equal(t, service.Name, "service")
+	assert.Equal(t, service.Namespace.Name, "namespace")
+	assert.Equal(t, service.Type, "ClusterIP")
+	assert.Equal(t, service.Ip, "fromservice")
+	assert.Equal(t, service.Labels, map[string]string{"label1": "labelName1", "label2": "labelName2"})
+	assert.Equal(t, service.Ports, models.Ports{
+		models.Port{Name: "http", Protocol: "TCP", Port: 3001},
+		models.Port{Name: "http", Protocol: "TCP", Port: 3000}})
+	assert.Equal(t, service.Endpoints, models.Endpoints{
+		models.Endpoint{
+			Addresses: models.Addresses{
+				models.Address{Kind: "Pod", Name: "recommendation-v1", IP: "172.17.0.9"},
+				models.Address{Kind: "Pod", Name: "recommendation-v2", IP: "172.17.0.8"},
+			},
+			Ports: models.Ports{
+				models.Port{Name: "http", Protocol: "TCP", Port: 3001},
+				models.Port{Name: "http", Protocol: "TCP", Port: 3000},
+			}}})
+	assert.Equal(t, service.Pods, models.Pods{
+		models.Pod{
+			Name: "Pod1",
+			Labels: map[string]string{
+				"label1": "labelName1Pod1",
+				"label2": "labelName2Pod1"}},
+		models.Pod{
+			Name: "Pod2",
+			Labels: map[string]string{
+				"label1": "labelName1Pod2",
+				"label2": "labelName2Pod2"}}})
+
+	// Istio Details
+	assert.Equal(t, service.RouteRules, models.RouteRules{
+		models.RouteRule{
+			Destination: "destination1",
+			Precedence:  "precedence1",
+			Route:       "route1"},
+		models.RouteRule{
+			Destination: "destination2",
+			Precedence:  "precedence2",
+			Route:       "route2"}})
+
+	// Prometheus Client
+	assert.Equal(t, service.Dependencies, map[string][]string{
+		"v1": []string{"unknown", "/products", "/reviews"},
+		"v2": []string{"/catalog", "/shares"}})
+}
+
+func fakeServiceDetails() *kubernetes.ServiceDetails {
+	service := &v1.Service{
```

Comment:
could (or should?) this be extracted to a config file (say json) rather than coded? then it could be used in several places in code

Reply:
I guess all opinions are valid, personally I like to have all defined in the same place for tests so I don't have to juggle with different files while reading, understanding or debugging a test

Reply:
@jotak  my thought was that we are going to have a lot of tests which often might rely on the same data (several services, some route rules, etc). So if we externalize that to some json file that is adjacent to tests, that can help not writing the same "create service...." every time in a new test. I'm open to other suggestions.

Reply:
@abonas I'm fine with externalization with one suggestion though: instead of json, directly build model using go structs (same as here). So that we spot faster any breaking change in the model, with compile errors rather than test failures.

Reply:
@jotak no problem, sounds good to me. I suggested a json as it was *a* way to do it, not necessarily *the* way :)